File: leet/0290_word_pattern/0290_word_pattern.js
// LeetCode 290. Word Pattern
// https://leetcode.com/problems/word-pattern/
// O(n) time-and-space
var wordPattern = function(pattern, s) {
    words = s.split(' ');
    if (pattern.length !== words.length)
        return false;
    const wordToLetter = new Map(), seenLetters = new Set();
    for (let i = 0; i < words.length; i++) {
        if (wordToLetter.has(words[i])) {
            if (wordToLetter.get(words[i]) !== pattern[i])
                return false;
        } else {
            if (seenLetters.has(pattern[i]))
                return false;
            wordToLetter.set(words[i], pattern[i]);
            seenLetters.add(pattern[i]);
        }
    }
    return true;
};

File: leet/0290_word_pattern/0290_word_pattern.py
# LeetCode 290. Word Pattern
# O(n) time-and-space
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        char_to_word = {}
        word_to_char = {}
        words = str.split(" ")
        if len(pattern) != len(words):
            return False
        for char, word in zip (list(pattern), words):
            if char not in char_to_word:
                if word in word_to_char:
                    return False
                char_to_word[char] = word
                word_to_char[word] = char
            else:
                if word not in word_to_char or char_to_word[char] != word:
                    return False
        return True
